### PR TITLE
Repo v2

### PR DESCRIPTION
**Pull request by dholms**
A few changes that we're getting in here:
- dropping the average fanout of the MST from 16 -> 4 (two lead `0`s per layer) for smaller proof sizes
- merging together the repo root, meta & commit into a single node
  - the `sig` on the commit is now the signature of the sha256 of the cbor-encoded commit (excluding the signature). This is the same method as PLC operations as well as git commits
  - removed `datastore` meta value
- mst keys are now cbor-encoded as `bytes` instead of `string`
- we've also imposed restrictions on mst keys
  - 0 < length <= 256
  - alphanumeric along with `-`, `_`, `.` & `:`
  - enforced collection semantics (`collection/rkey`). That is _exactly_ one `/` with some characters on either side
- and of course 🥁 🥁 bumping the version to 2!
  - we won't be keeping around code paths for version 1, but there are repos out in the wild & it will be nice to be able to determine which version they are

I'll be making a followup PR with the migration that goes along with this

### CODE REVIEW

File: packages/repo/tests/mst.test.ts [hunk 1]
```diff
@@ -155,378 +155,273 @@ describe('Merkle Search Tree', () => {
     }
   })
 
-  // Special Cases (these are made for fanout 32)
-  // ------------
```
Comment (devinivy) [inline packages/repo/tests/mst.test.ts:159]: I know these special cases were specific to the fanout— will we end-up recreate new versions of them for the new fanout?

**Review by devinivy — APPROVED** [review]
Feels real good!  Wasn't able to spot anything of serious note.

File: packages/pds/tests/sync/subscribe-all-repos.test.ts [hunk 2]
```diff
@@ -104,7 +104,7 @@ describe('repo subscribe all repos', () => {
         cid: w.action === WriteOpAction.Delete ? null : w.cid.toString(),
       }))
       const sortedOps = evt.ops.sort((a, b) => a.path.localeCompare(b.path))
-      const sortedWrites = evt.ops.sort((a, b) => a.path.localeCompare(b.path))
+      const sortedWrites = writes.sort((a, b) => a.path.localeCompare(b.path))
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/pds/tests/sync/subscribe-all-repos.test.ts:107]: fortunately the test still passed fine 😅

File: packages/repo/src/mst/mst.ts [hunk 3]
```diff
@@ -52,8 +46,8 @@ import * as parse from '../parse'
  */
 const subTreePointer = z.nullable(common.cid)
 const treeEntry = z.object({
-  p: z.number(), // prefix count of utf-8 chars that this key shares with the prev key
-  k: z.string(), // the rest of the key outside the shared prefix
+  p: z.number(), // prefix count of ascii chars that this key shares with the prev key
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/mst/mst.ts:49]: yup that's right 👍

File: packages/repo/src/mst/util.ts [hunk 4]
```diff
@@ -120,3 +113,27 @@ export const cidForEntries = async (entries: NodeEntry[]): Promise<CID> => {
...
+export const validCharsRegex = /^[a-zA-Z0-9_\-:.]*$/
+
+export const isValidMstKey = (str: string): boolean => {
+  if (str.length >= 256) return false
+  const split = str.split('/')
+  if (split.length !== 2) return false
+  if (split[0].length < 1) return false
+  if (split[1].length < 1) return false
+  if (!validCharsRegex.test(split[0])) return false
+  return validCharsRegex.test(split[1])
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/mst/util.ts:126]: oh nice I dig 👍

File: packages/repo/src/types.ts [hunk 5]
```diff
@@ -7,44 +7,30 @@ import BlockMap from './block-map'
...
-export type RepoMeta = z.infer<typeof repoMeta>
-
-const repoRoot = z.object({
-  meta: common.cid,
   prev: common.cid.nullable(),
-  auth_token: common.cid.nullable().optional(),
   data: common.cid,
 })
-export type RepoRoot = z.infer<typeof repoRoot>
+export type UnsignedCommit = z.infer<typeof unsignedCommit>
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/types.ts:16]: ah good call. that actually bit me once already 😅

File: packages/repo/src/verify.ts [hunk 6]
```diff
@@ -100,49 +92,46 @@ export const verifyUpdates = async (
...
+  did: string,
+  signingKey: string,
 ): Promise<VerifiedUpdate[]> => {
   const commitPath = await updateStorage.getCommitPath(updateRoot, repo.cid)
   if (commitPath === null) {
     throw new RepoVerificationError('Could not find shared history')
   }
   const syncStorage = new SyncStorage(updateStorage, repo.storage)
-  return verifyCommitPath(repo, syncStorage, commitPath, didResolver)
+  return verifyCommitPath(repo, syncStorage, commitPath, did, signingKey)
```
[the remark this answers is not in the document]
Comment (dholms) [reply packages/repo/src/verify.ts:103]: yup absolutely. plus this lets the consumer decide how they want to deal with did resolution & out of date dids, etc. less magic ✨

File: packages/repo/tests/mst.test.ts (hunk 1 above)
Comment (dholms) [reply packages/repo/tests/mst.test.ts:159]: Yup the interop stuff at the bottom of the test file is the exact same tests but targeted for fanout 4 👍

File: packages/repo/src/mst/util.ts [hunk 7]
```diff
@@ -120,3 +113,33 @@ export const cidForEntries = async (entries: NodeEntry[]): Promise<CID> => {
   const data = serializeNodeData(entries)
   return cidForCbor(data)
 }
+
+export const isValidMstKey = (str: string): boolean => {
+  const split = str.split('/')
+  return (
+    str.length <= 256 &&
```
Comment (bnewbold) [inline packages/repo/src/mst/util.ts:120]: I *think* that this length check works ok. It is technically the number of bytes that should be <= 256, and a typescript string is UTF-16 and could be a different number of bytes. But I can't think of any evil strings off the top of my head where this code, with the `split('/')` and then double regex check, would ever have the "string" length be different from the "bytes" length.

**Comment by bnewbold** [comment]
I am pretty excited about this! Consolidating the commit record is a big simplicity win. Feels like MST/repo will be much more resistant to misuse, either unintentional or :trollface:.

It should be straightforward to port these changes to golang and rust. In particular there was almost no structural change so the golang implementation remains 1-to-1 with this typescript implementation.

Clearly documenting the signing processes will be important. Eg, signature of the SHA-256 as hex-encoded string vs bytes. It does feel a bit like generating a CID for the commit object and signing the CID could be easier for some implementations compared to doing DAG-CBOR encoding and SHA-256 and signing that. Just because there is usually a "IPLD object to CID" helper handy. But CID is also more complex and this is reasonable and already used in PLC, we just need to document it. And maybe add some more interop tests (in a followup), eg with a fixed/known signing private key.

**Review by devinivy — APPROVED** [review]

**Pull request merged**